source/common/network/listener_impl.cc

```diff
@@ -9,19 +9,53 @@
 #include "common/network/connection_impl.h"
 #include "common/ssl/connection_impl.h"
 
+#include "server/connection_handler.h"
+
 #include "event2/listener.h"
 
 namespace Network {
 
+void ListenerImpl::listenCallback(evconnlistener*, evutil_socket_t fd, sockaddr* addr, int,
+                                  void* arg) {
+  ListenerImpl* listener = static_cast<ListenerImpl*>(arg);
+
+  if (listener->use_original_dst_ && listener->connection_handler_ != nullptr) {
+    struct sockaddr_storage orig_dst_addr;
+    memset(&orig_dst_addr, 0, sizeof(orig_dst_addr));
+
+    bool success = Utility::getOriginalDst(fd, &orig_dst_addr);
+
+    if (success) {
+      std::string orig_sock_name = std::to_string(
+          listener->getAddressPort(reinterpret_cast<struct sockaddr*>(&orig_dst_addr)));
+
+      if (listener->socket_.name() != orig_sock_name) {
+        ListenerImpl* new_listener =
```

Comment:
The main reason why this is "needed" is that there is no abstract class for ConnectionHandler. If the presence of a ConnectionHandler is an implementation detail that only ListenerImpl should know about (not Listener) then I need to cast. If it is ok to expose ConnectionHandler as a first-class component, then it can be part of the listener interface and there's no need to cast. We can even make the pointer to ConnectionHandler part of the Listener constructor and get rid of the connectionHandler(ConnectionHandler* conn_handler) function which I didn't really like.